crates/m3/src/gadgets/barrel_shifter.rs

```diff
@@ -0,0 +1,252 @@
+	/// right shift).
+	flags: BarrelShifterFlags,
+}
+
+impl BarrelShifter {
+	/// Creates a new instance of the `BarrelShifter` gadget.
+	///
+	/// # Arguments
+	///
+	/// * `table` - A mutable reference to the `TableBuilder` used to define the
+	///   gadget.
+	/// * `flags` - A `BarrelShifterFlags` struct that configures the behavior
+	///   of the gadget.
+	///
+	/// # Returns
+	///
+	/// A new instance of the `BarrelShifter` gadget.
+	pub fn new(
+		table: &mut TableBuilder,
+		input: Col<B1, 32>,
+		shift_amount: Col<B1, 16>,
+		flags: BarrelShifterFlags,
+	) -> Self {
+		let partial_shift =
+			core::array::from_fn(|i| table.add_committed(format!("partial_shift_{i}")));
+		let shift_ammount_bits: [_; MAX_SHIFT_BITS] = core::array::from_fn(|i| {
+			table.add_selected(format!("shift_ammount_bits_{i}"), shift_amount, i)
+		});
+		let mut shifted = Vec::with_capacity(MAX_SHIFT_BITS);
+		let mut current_shift = input;
+		for i in 0..MAX_SHIFT_BITS {
+			shifted.push(table.add_shifted("shifted", current_shift, 5, 1 << i, flags.variant));
+			let partial_shift_packed: Col<B32> =
+				table.add_packed(format!("partial_shift_packed_{i}"), partial_shift[i]);
+			let shifted_packed: Expr<B32, 1> = table
+				.add_packed(format!("shifted_packed_{i}"), shifted[i])
+				.into();
+			let current_shift_packed: Col<B32> =
+				table.add_packed(format!("current_shift_packed_{i}"), current_shift);
+			table.assert_zero(
+				format!("correct_partial_shift_{i}"),
+				partial_shift_packed
+					- (shifted_packed * upcast_col(shift_ammount_bits[i])
+						+ current_shift_packed * (upcast_col(shift_ammount_bits[i]) + B32::ONE)),
+			);
+			current_shift = partial_shift[i];
+		}
+
+		// Define the output column (32 bits).
+		let output = if flags.commit_output {
+			// If the output is committed, add a committed column and enforce constraints.
+			let output = table.add_committed::<B1, 32>("output");
+			table.assert_zero("output_constraint", current_shift - output);
+			output
+		} else {
+			current_shift
+		};
+
+		Self {
+			input,
+			shift_amount,
+			shift_ammount_bits,
+			shifted,
+			partial_shift,
+			output,
+			flags,
+		}
+	}
+
+	/// Populates the table with witness values for the barrel shifter.
+	///
+	/// # Arguments
+	///
+	/// * `witness` - A mutable reference to the `TableWitness` used to populate
+	///   the table.
+	///
+	/// # Returns
+	///
+	/// A `Result` indicating success or failure.
+	pub fn populate<P>(&self, index: &mut TableWitnessSegment<P>) -> Result<(), anyhow::Error>
+	where
+		P: PackedFieldIndexable<Scalar = B128> + PackedExtension<B1>,
+	{
+		let input: RefMut<'_, [u32]> = index.get_mut_as(self.input).unwrap();
+		let shift_ammount: RefMut<'_, [u16]> = index.get_mut_as(self.shift_amount).unwrap();
+		// TODO: Propagate the errors
+		let mut partial_shift: [_; MAX_SHIFT_BITS] =
+			core::array::from_fn(|i| index.get_mut_as(self.partial_shift[i]).unwrap());
```

Comment:
You can use `array_util::try_from_fn`

Reply:
Done, but now I'm getting
`[build-debug-amd-stable: crates/m3/src/lib.rs#L80](https://github.com/IrreducibleOSS/binius/pull/232/files#annotation_33925323140)
`#![feature]` may not be used on the stable release channel`

Reply:
oh... no, I would assume jimpo meant non-std version of the function, i.e. [this one](https://docs.rs/array-util/latest/array_util/fn.try_from_fn.html).